core/txpool/legacypool/legacypool.go

```diff
@@ -883,6 +883,16 @@ func (pool *LegacyPool) promoteTx(addr common.Address, hash common.Hash, tx *typ
 		pendingDiscardMeter.Mark(1)
 		return false
 	}
+
+	//Check balance for Overdraft
+	balance := pool.currentState.GetBalance(addr)
+	if list.totalcost.Cmp(balance) > 0 {
+		pool.all.Remove(hash)
+		pool.priced.Removed(1)
+		pendingDiscardMeter.Mark(1)
+		return false
+	}
```

Comment:
Intuitively I think this check should be done before the tx is added to the list. Feels unnecessary to add it to the list simply to calculate the total cost. Another possibility would be to add a new `Add` method (or change the current signature?) to pass in the accounts current balance so it can decide for itself if it should add the tx to this.

Also the current implementation adds removes the tx from the pool, but it doesn't remove it from the accounts pending list.